src/strands/multiagent/graph.py

```diff
@@ -230,44 +323,75 @@ def build(self) -> "Graph":
         invalid_entries = entry_point_ids - set(self.nodes.keys())
         if invalid_entries:
             raise ValueError(f"Entry points not found in nodes: {invalid_entries}")
 
-        # Check for cycles using DFS with color coding
-        WHITE, GRAY, BLACK = 0, 1, 2
-        colors = {node_id: WHITE for node_id in self.nodes}
-
-        def has_cycle_from(node_id: str) -> bool:
-            if colors[node_id] == GRAY:
-                return True  # Back edge found - cycle detected
-            if colors[node_id] == BLACK:
+        # Check for cycles only if not explicitly allowed
+        if not self._allow_cycles:
+            # Check for cycles using DFS with color coding
+            WHITE, GRAY, BLACK = 0, 1, 2
+            colors = {node_id: WHITE for node_id in self.nodes}
+
+            def has_cycle_from(node_id: str) -> bool:
+                if colors[node_id] == GRAY:
+                    return True  # Back edge found - cycle detected
+                if colors[node_id] == BLACK:
+                    return False
+
+                colors[node_id] = GRAY
+                # Check all outgoing edges for cycles
+                for edge in self.edges:
+                    if edge.from_node.node_id == node_id and has_cycle_from(edge.to_node.node_id):
+                        return True
+                colors[node_id] = BLACK
                 return False
 
-            colors[node_id] = GRAY
-            # Check all outgoing edges for cycles
-            for edge in self.edges:
-                if edge.from_node.node_id == node_id and has_cycle_from(edge.to_node.node_id):
-                    return True
-            colors[node_id] = BLACK
-            return False
-
-        # Check for cycles from each unvisited node
-        if any(colors[node_id] == WHITE and has_cycle_from(node_id) for node_id in self.nodes):
-            raise ValueError("Graph contains cycles - must be a directed acyclic graph")
+            # Check for cycles from each unvisited node
+            if any(colors[node_id] == WHITE and has_cycle_from(node_id) for node_id in self.nodes):
+                raise ValueError("Graph contains cycles - use allow_cycles() to enable cyclic graphs")
+        elif self._max_node_executions is None and self._execution_timeout is None:
+            logger.warning("Cyclic graphs without limits may run indefinitely")
 
 
 class Graph(MultiAgentBase):
-    """Directed Acyclic Graph multi-agent orchestration."""
+    """Directed Graph multi-agent orchestration with optional cycle support."""
 
-    def __init__(self, nodes: dict[str, GraphNode], edges: set[GraphEdge], entry_points: set[GraphNode]) -> None:
-        """Initialize Graph."""
+    def __init__(
+        self,
+        nodes: dict[str, GraphNode],
+        edges: set[GraphEdge],
+        entry_points: set[GraphNode],
+        max_node_executions: int | None = None,
+        execution_timeout: float | None = None,
+        node_timeout: float | None = None,
```

Comment:
nit: Instead of `int | None`, we typically use Optional
```suggestion
        max_node_executions: Optional[int] = None,
        execution_timeout: Optional[float] = None,
        node_timeout: Optional[float] = None,
```